---
title: "Single Cell RNA-Seq Preprocessing Pipeline"
author: "Alexander G Foote"
date: "`r Sys.Date()`"
output:
    html_document:
      keep_md: TRUE
      toc: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, error = FALSE)
```

## Set up Workspace
```{r libraries, include=FALSE}
library(Seurat)     # single cell RNA-Seq analysis
library(kableExtra) # format tables
library(ggplot2)  
library(biomaRt)
library(tidyr) # create graphics
library(tidyverse)
library(viridis)   # accessible color palettes
library(hdf5r)
library(DoubletFinder)
library(patchwork)
library(BiocManager)
library(gprofiler2)
set.seed(12345)
```

## Create object

The `CreateSeuratObject` function allows feature (gene) and cell filtering by minimum cell and feature counts. We will set these to 0 for now in order to explore manual filtering more fully later on.
```{r create_seurat_objects}
saline1.data <- Read10X_h5("/Users/alexanderfoote/raw_data/Cell Ranger Matix Data Files/IAV_mouse-genome_cellranger_02142024_7dp-IAV/control_v1/Crtl1_filtered_feature_bc_matrix.h5", use.names = TRUE, unique.features = TRUE)
scRNA_saline1_7dpi="all_compartments_saline1"
saline1 <- CreateSeuratObject(counts = saline1.data, project = scRNA_saline1_7dpi, min.cells = 0, min.features = 0, names.field = 2,  names.delim = "\\_")
class(saline1[["RNA"]])
head(saline1@meta.data)

saline2.data <- Read10X_h5('/Users/alexanderfoote/raw_data/Cell Ranger Matix Data Files/IAV_mouse-genome_cellranger_02142024_7dp-IAV/control_v2/Crtl2_filtered_feature_bc_matrix.h5', use.names = TRUE, unique.features = TRUE)
scRNA_saline2_7dpi="all_compartments_saline2"
saline2 <- CreateSeuratObject(counts = saline2.data, project = scRNA_saline2_7dpi, min.cells = 0, min.features = 0, names.field = 2,  names.delim = "\\_")
class(saline2[["RNA"]])
head(saline2@meta.data)

IAV.data <- Read10X_h5('/Users/alexanderfoote/raw_data/Cell Ranger Matix Data Files/IAV_mouse-genome_cellranger_02142024_7dp-IAV/flu/IAV_filtered_feature_bc_matrix.h5', use.names = TRUE, unique.features = TRUE)
scRNA_IAV_7dpi="all_compartments_IAV"
IAV <- CreateSeuratObject(counts = IAV.data, project = scRNA_IAV_7dpi, min.cells = 0, min.features = 0, names.field = 2,  names.delim = "\\_")
class(IAV[["RNA"]])
head(IAV@meta.data)
```

### Mitochondrial gene expression

Filtering on the expression of genes from the mitochondrial genome is not appropriate in all cell types. However, in many tissues, low-quality / dying cells may exhibit extensive mitochondrial contamination. Even when not filtering on mitochondrial expression, the data can be interesting or informative. The `PercentageFeatureSet` function calculates the proportion of counts originating from a set of features. Genes in the human mitochondrial genome begin with 'MT', while those in the mouse mitochondrial genome begin with 'mt'.
```{r percent_MT}
saline1[["percent.mt"]] <- PercentageFeatureSet(saline1, pattern = "^mt-") 
head(saline1@meta.data)
saline2[["percent.mt"]] <- PercentageFeatureSet(saline2, pattern = "^mt-") 
head(saline2@meta.data)
IAV[["percent.mt"]] <- PercentageFeatureSet(IAV, pattern = "^mt-") 
head(IAV@meta.data) 
```

### Visualize distribution of metadata values

The `VlnPlot` function produces a composite plot with one panel for each element of the "features" vector. The data are grouped by the provided identity; by default, this is the active identity of the object, which can be accessed using the `Idents()` function, or in the "active.ident" slot.
```{r violins, fig.height=12}
VlnPlot(saline1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(saline2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(IAV, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

### Scatter plots
```{r relationships}
# saline1
FeatureScatter(saline1, feature1 = "nCount_RNA", feature2 = "percent.mt",  pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 60000)) +
  geom_hline(yintercept = 15) 
FeatureScatter(saline1, feature1 = "nFeature_RNA", feature2 = "percent.mt", pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 7500)) +
  geom_hline(yintercept = 15) 
FeatureScatter(saline1, feature1 = "nFeature_RNA", feature2 = "nCount_RNA", pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 7500)) +
  geom_hline(yintercept = 40000) 
# saline2
FeatureScatter(saline2, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 60000)) +
  geom_hline(yintercept = 15) 
FeatureScatter(saline2, feature1 = "nFeature_RNA", feature2 = "percent.mt", pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 7500)) +
  geom_hline(yintercept = 15) 
FeatureScatter(saline2, feature1 = "nFeature_RNA", feature2 = "nCount_RNA", pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 7500)) +
  geom_hline(yintercept = 40000) 
# IAV
FeatureScatter(IAV, feature1 = "nCount_RNA", feature2 = "percent.mt", pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 60000)) +
  geom_hline(yintercept = 15) 
FeatureScatter(IAV, feature1 = "nFeature_RNA", feature2 = "percent.mt", pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 7500)) +
  geom_hline(yintercept = 15) 
FeatureScatter(IAV, feature1 = "nFeature_RNA", feature2 = "nCount_RNA", pt.size = 0.5, shuffle = TRUE) +
  geom_vline(xintercept = c(200, 7500)) +
  geom_hline(yintercept = 40000) 
```

## Cell filtering/QC

The goal of cell filtering is to remove cells with anomolous expression profiles, typically low UMI cells, which may correspond to low-quality cells or background barcodes that made it through the Cell Ranger filtration algorithm. It may also be appropriate to remove outlier cells with extremely high UMI counts. In this case, the proposed cut-offs on the high end of the distributions are quite conservative, in part to reduce the size of the object and speed up analysis during the workshop. These filters can be put in place with the `subset` function.
```{r, cell_filtering}
# Pre-check
saline1
table(saline1$orig.ident)
saline2
table(saline2$orig.ident)
IAV
table(IAV$orig.ident)
#Filtering objects
saline1.filter <- subset(saline1, subset = nFeature_RNA >= 200 & nFeature_RNA <= 7500 & nCount_RNA >= 200 & nCount_RNA <= 60000 & percent.mt < 15)
saline1 <- saline1.filter
saline2.filter <- subset(saline2, subset = nFeature_RNA >= 200 & nFeature_RNA <= 7500 & nCount_RNA >= 200 & nCount_RNA <= 60000 & percent.mt < 15)
saline2 <- saline2.filter
IAV.filter <- subset(IAV, subset = nFeature_RNA >= 200 & nFeature_RNA <= 7500 & nCount_RNA >= 200 & nCount_RNA <= 60000 & percent.mt < 15)
IAV <- IAV.filter
# Post-check
saline1
table(saline1$orig.ident)
saline2
table(saline2$orig.ident)
IAV
table(IAV$orig.ident)
```

## Feature filtering

When creating the base Seurat object, we had the opportunity filter out some genes using the "min.cells" argument. At the time, we set that to 0. Since we didn't filter our features then, we can apply a filter at this point. If we had filtered when the object was created, this would be an opportunity to be more aggressive. The custom code below provides a function that filters genes requiring a min.umi in at least min.cells, or takes a user-provided list of genes.
```{r gene_filter}
# define function
FilterGenes <- function(object, min.umi = NA, min.cells = NA, genes = NULL) {
  genes.use = NA
  if (!is.null(genes)) {
    genes.use = intersect(rownames(object), genes)
    } else if (min.cells & min.umi) {
      num.cells = Matrix::rowSums(GetAssayData(object) >= min.umi)
      genes.use = names(num.cells[which(num.cells >= min.cells)])
    }
  object = object[genes.use,]
  object = LogSeuratCommand(object = object)
  return(object)
}
# apply filter 
saline1.filter <- FilterGenes(object = saline1, min.umi = 2, min.cells = 10)
saline2.filter <- FilterGenes(object = saline2, min.umi = 2, min.cells = 10)
IAV.filter <- FilterGenes(object = IAV, min.umi = 2, min.cells = 10)
# filtering results 
saline1
saline1 <- saline1.filter
saline2
saline2 <- saline2.filter
IAV
IAV <- IAV.filter
```

## Normalize data

After filtering, the next step is to normalize the data. We employ a global-scaling normalization method, LogNormalize, that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and then log-transforms the data.
```{r normalize}
saline1 <- NormalizeData(object = saline1, normalization.method = "LogNormalize", scale.factor = 10000)
saline2 <- NormalizeData(object = saline2, normalization.method = "LogNormalize", scale.factor = 10000)
IAV <- NormalizeData(object = IAV, normalization.method = "LogNormalize", scale.factor = 10000)
```

## Cell cycle assignment

Cell cycle phase can be a significant source of variation in single cell and single nucleus experiments. There are a number of automated cell cycle stage detection methods available for single cell data. For this workshop, we will be using the built-in Seurat cell cycle function, `CellCycleScoring`. This tool compares gene expression in each cell to a list of cell cycle marker genes and scores each barcode based on marker expression. The phase with the highest score is selected for each barcode. Seurat includes a list of cell cycle genes in human single cell data.
```{r human_gene_list}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
```

For other species, a user-provided gene list may be substituted, or the orthologs of the human gene list used instead.

**Do not run the code below for human experiments!**
```{r convert_list_mouse, eval=FALSE}
# mouse code DO NOT RUN for human data
convertHumanGeneList <- function(x){
  require("biomaRt")
  human = useEnsembl("ensembl",
                     dataset = "hsapiens_gene_ensembl",
                     mirror = "uswest")
  mouse = useEnsembl("ensembl",
                     dataset = "mmusculus_gene_ensembl",
                     mirror = "uswest")
  genes = getLDS(attributes = c("hgnc_symbol"),
                 filters = "hgnc_symbol",
                 values = x ,
                 mart = human,
                 attributesL = c("mgi_symbol"),
                 martL = mouse,
                 uniqueRows=T)
  humanx = unique(genes[, 2])
  print(head(humanx)) # print first 6 genes found to the screen
  return(humanx)
}
# convert lists to mouse orthologs
s.genes = gorth(cc.genes.updated.2019$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
g2m.genes = gorth(cc.genes.updated.2019$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
```

Once an appropriate gene list has been identified, the `CellCycleScoring` function can be run.
```{r CellCycleScoring}
saline1 <- CellCycleScoring(saline1,
                            s.features = s.genes,
                            g2m.features = g2m.genes,
                            set.ident = TRUE)
table(saline1@meta.data$Phase) %>%
  kable(table.attr = "style = \"color: black;\"", caption = "# Cells in each Cell Cycle Stage_Saline1",
        col.names = c("Stage", "Count"),
        align = "c") %>%
  kable_styling()
saline2 <- CellCycleScoring(saline2,
                            s.features = s.genes,
                            g2m.features = g2m.genes,
                            set.ident = TRUE)
table(saline2@meta.data$Phase) %>%
  kable(table.attr = "style = \"color: black;\"", caption = "# Cells in each Cell Cycle Stage_Saline2",
        col.names = c("Stage", "Count"),
        align = "c") %>%
  kable_styling()
IAV <- CellCycleScoring(IAV,
                            s.features = s.genes,
                            g2m.features = g2m.genes,
                            set.ident = TRUE)
table(IAV@meta.data$Phase) %>%
  kable(table.attr = "style = \"color: black;\"", caption = "# Cells in each Cell Cycle Stage_IAV",
        col.names = c("Stage", "Count"),
        align = "c") %>%
  kable_styling()
```

## Identify variable genes

The function FindVariableFeatures identifies the most highly variable genes (default 2000 genes) by fitting a line to the relationship of log(variance) and log(mean) using loess smoothing, uses this information to standardize the data, then calculates the variance of the standardized data. This helps avoid selecting genes that only appear variable due to their expression level.
```{r find_variable_genes}
Idents(saline1) <- "orig.ident"
Idents(saline2) <- "orig.ident"
Idents(IAV) <- "orig.ident"

saline1 <- FindVariableFeatures(object = saline1, selection.method = "vst")
length(VariableFeatures(saline1))

top10 <- head(VariableFeatures(saline1), 10)
top10

var.feat.plot <- VariableFeaturePlot(saline1)
var.feat.plot <- LabelPoints(plot = var.feat.plot, points = top10, repel = TRUE)
var.feat.plot

saline2 <- FindVariableFeatures(object = saline2, selection.method = "vst")
length(VariableFeatures(saline2))

top10 <- head(VariableFeatures(saline2), 10)
top10

var.feat.plot <- VariableFeaturePlot(saline2)
var.feat.plot <- LabelPoints(plot = var.feat.plot, points = top10, repel = TRUE)
var.feat.plot

IAV <- FindVariableFeatures(object = IAV, selection.method = "vst")
length(VariableFeatures(IAV))

top10 <- head(VariableFeatures(IAV), 10)
top10

var.feat.plot <- VariableFeaturePlot(IAV)
var.feat.plot <- LabelPoints(plot = var.feat.plot, points = top10, repel = TRUE)
var.feat.plot
```

## Scale the data 

**This step is computationally intensive and can be quite slow.**

The `ScaleData` function scales and centers genes in the dataset. If variables are provided with the "vars.to.regress" argument, they are individually regressed against each gene, and the resulting residuals are then scaled and centered unless otherwise specified. We regress out cell cycle results S.Score and G2M.Score, mitochondrial RNA level (percent_mt), and the number of features (nFeature_RNA) as a proxy for sequencing depth.
```{r ScaleData}
saline1 <- ScaleData(saline1, vars.to.regress = c("S.Score", "G2M.Score", "percent.mt", "nFeature_RNA"))
saline2 <- ScaleData(saline2, vars.to.regress = c("S.Score", "G2M.Score", "percent.mt", "nFeature_RNA"))
IAV <- ScaleData(IAV, vars.to.regress = c("S.Score", "G2M.Score", "percent.mt", "nFeature_RNA"))
```

## Perform dimensionality reduction
```{r pca}
saline1 <- RunPCA(saline1, npcs = 100)
saline2 <- RunPCA(saline2, npcs = 100)
IAV <- RunPCA(IAV, npcs = 100)
```

While it is theoretically possible to calculate as many PCs as there are features in the data, typically 100 PCs is more than sufficient. In fact, many of these PCs may explain negligible amounts of variation. Seurat provides a number of ways to visualize the PCA results.

### Principal components plot

The PCA biplot is a scatter plot showing the placement of each cell on two selected components. By default, the first and second PC are used, but any two calculated PCs may be specified. At this point in the analysis, since we are no longer performing QA and filtering, we can move to examining relationships between cells on a per-group rather than per-sample basis.

```{r plot_pca}
DimPlot(saline1, reduction = "pca", dims = c(1,2), shuffle = TRUE)
DimPlot(saline2, reduction = "pca", dims = c(1,2), shuffle = TRUE) 
DimPlot(IAV, reduction = "pca", dims = c(1,2), shuffle = TRUE)
```

The axes are unit-less; points (cells or nuclei) that are farther apart are more dissimilar on the displayed PC than points that are closer together.

### PCA loadings

Each PC can be imagined as a sort of meta-gene for which every cell has an expression value. The top genes associated with the reduction component (i.e. contributing to a cell's "expression level" of that meta-gene) can be plotted for a selected dimension(s) using the `VizDimLoadings` function.
```{r viz_pca}
VizDimLoadings(saline1, dims = 1:2, nfeatures = 25, reduction = "pca", ncol = 2) + theme_minimal(base_size = 8)

VizDimLoadings(saline2, dims = 1:2, nfeatures = 25, reduction = "pca", ncol = 2) + theme_minimal(base_size = 8)

VizDimLoadings(IAV, dims = 1:2, nfeatures = 25, reduction = "pca", ncol = 2) + theme_minimal(base_size = 8)
```

### Heat map

Heat maps display similar information. On the x-axis, cells are ordered by their embeddings ("expression" of the PC), while on the y-axis, genes are ordered by PC loading. When fewer than the total number of cells is selected, this results in selection of the cells with the largest absolute value embeddings, which emphasizes variation on the PC.
```{r heatmap_pca}
DimHeatmap(saline1, dims = 1:2, nfeatures = 25, cells = 500, reduction = "pca", balanced = TRUE, slot = "scale.data")
DimHeatmap(saline2, dims = 1:2, nfeatures = 25, cells = 500, reduction = "pca", balanced = TRUE, slot = "scale.data")
DimHeatmap(IAV, dims = 1:2, nfeatures = 25, cells = 500, reduction = "pca", balanced = TRUE, slot = "scale.data")
```

## Selecting PCs to use

To overcome the extensive technical noise in any single gene, Seurat clusters cells based on their PCA scores, with each PC essentially representing a meta-gene that combines information across a correlated gene set. Determining how many PCs to include downstream is therefore an important step.

### Elbow plot

An elbow plot displays the standard deviations (or approximate singular values if running PCAFast) of the principle components for easy identification of an elbow in the graph. This elbow often corresponds well with the significant PCs and is much faster to run. This is the traditional approach to selecting principal components. The appearance of elbow plots tends to be highly consistent across single cell / single nucleus experiments. Generally, the line approaches zero at around PC 50. This is a reasonable number of PCs to use for the downstream steps.
```{r elbow}
ElbowPlot(saline1, ndims = 50)
ElbowPlot(saline2, ndims = 50)
ElbowPlot(IAV, ndims = 50)
```

PCs with a strong enrichment of low p-value genes are identified as significant components. Let's use the first 50 PCs. 50 PCs is quite standard for most all datasets and is widely used as a chosen PC score.

## UMAP [Uniform Manifold Approximation and Projection](https://arxiv.org/pdf/1802.03426v3.pdf) 

UMAP is a dimensionality reduction method that is commonly used in single cell RNA-Seq analysis. Single cell data is extremely high-dimensional; UMAP calculates a nearest neighbor network describing the relationships between cells as measured by the PC loadings of variable genes and creates a low-dimensional space that preserves these relationships.
```{r UMAP}
saline1 <- RunUMAP(saline1, dims = 1:50)
saline1 <- FindNeighbors(saline1, dims = 1:50)
saline1 <- FindClusters(saline1, resolution = 0.5) 

saline2 <- RunUMAP(saline2, dims = 1:50)
saline2 <- FindNeighbors(saline2, dims = 1:50)
saline2 <- FindClusters(saline2, resolution = 0.5) 

IAV <- RunUMAP(IAV, dims = 1:50)
IAV <- FindNeighbors(IAV, dims = 1:50)
IAV <- FindClusters(IAV, resolution = 0.5) 
```

```{r}
# UMAP colored by sample identity
DimPlot(saline1, reduction = "umap", shuffle = TRUE) 
colnames(saline1@meta.data)
DimPlot(saline1, group.by = "Phase", reduction = "umap", shuffle = TRUE) +
  scale_color_viridis_d()
FeaturePlot(saline1, features = "G2M.Score", reduction = "umap")

DimPlot(saline2, reduction = "umap", shuffle = TRUE) 
colnames(saline2@meta.data)
DimPlot(saline2, group.by = "Phase", reduction = "umap", shuffle = TRUE) +
  scale_color_viridis_d()
FeaturePlot(saline2, features = "G2M.Score", reduction = "umap")

DimPlot(IAV, reduction = "umap", shuffle = TRUE) 
colnames(IAV@meta.data)
DimPlot(IAV, group.by = "Phase", reduction = "umap", shuffle = TRUE) +
  scale_color_viridis_d()
FeaturePlot(IAV, features = "G2M.Score", reduction = "umap")
```

## Doublet Detection/Removal

Doublets are cells that appear to be, but are not, real cells. There are two major types of doublets: heterotypic and homotypic. *Heterotypic doublets* are formed by cells with distinct transcriptional profiles. *Homotypic doublets* are formed by cells with similar transcriptional profiles. Heterotypic doublets are relatively easier to detect compared with homotypic doublets. Depending on the protocols used to barcode single cells/nuclei, doublet rates vary significantly and it can reach as high as 40%.

[DoubletFinder](https://github.com/chris-mcginnis-ucsf/DoubletFinder) takes fully pre-processed data from Seurat (NormalizeData, FindVariableGenes, ScaleData, RunPCA, and RunUMAP) as input and the process should be done for each sample individually. The input data should be processed to remove low-quality cell clusters first.

## Parameter selection
In addition to the Seurat object, DoubletFinder takes a number of arguments. These are:
* PCs: a vector of statistically significant PCs to use
* pN: the number of artificially generated doublets (default = 0.25)
* pK: PC neighborhood size used to compute network
* nExp: threshold used to make doublet/singlet call

We will use PCs 1-50 (based on the dimensionality reduction section).

The optimal value of pK varies between samples, and is impacted by the number of cell states and magnitude of transcriptional heterogeneity present in the data. The mean-variance normalized bimodality coefficient (BCmvn) is used as a measure of pK optimization. In experiments with known doublet frequencies, BCmvn is maximized by values of pK that produce the most accurate doublet-calling. In the code below, the pN & pK values corresponding to the maxiumum BCmvn are selected for each sample.
```{r include=FALSE}
# This increases the max allowed size for globals to 1 GB. Adjust the value if needed
options(future.globals.maxSize = 1000 * 1024^2)  # sets limit to 1000 MB (1GB)

#SALINE1
## pK Identification
sweep.res.list.saline1  <- paramSweep(saline1,PCs=1:50,sct=TRUE)
sweep.stats.saline1 <- summarizeSweep(sweep.res.list.saline1,GT=FALSE)
bcmvn.saline1 <- find.pK(sweep.stats.saline1)
sweep.stats.saline1 #use highest BCrea1 value here to chose the pN and pK
saline1
nExp_poi <- round(0.02880*nrow(saline1@meta.data)) #0.028 represents ~2800 total cells in condition (i.e. samples)

# Run DoubletFinder with chosen parameters
saline1 <- doubletFinder(saline1, PCs=1:50, pN=0.1, pK=0.01, nExp=nExp_poi, reuse.pANN=FALSE, sct=TRUE)

# Identify the doublet classification column (its name includes the parameters used)
df_col <- grep("DF.classifications", colnames(saline1@meta.data), value = TRUE)
cat("DoubletFinder classification column:", df_col, "\n")

# Subset the Seurat object to keep only cells classified as "Singlet"
saline1 <- subset(saline1, subset = DF.classifications_0.1_0.01_83 == "Singlet")
head(saline1@meta.data)

#SALINE2
sweep.res.list.saline2  <- paramSweep(saline2,PCs=1:50,sct=TRUE)
sweep.stats.saline2 <- summarizeSweep(sweep.res.list.saline2,GT=FALSE)
bcmvn.saline2 <- find.pK(sweep.stats.saline2)
sweep.stats.saline2
saline2
nExp_poi <- round(0.05835*nrow(saline2@meta.data)) 

# Run DoubletFinder with chosen parameters
saline2 <- doubletFinder(saline2, PCs=1:50, pN=0.15, pK=0.005, nExp=nExp_poi, reuse.pANN=FALSE, sct=TRUE) 

# Identify the doublet classification column (its name includes the parameters used)
df_col <- grep("DF.classifications", colnames(saline2@meta.data), value = TRUE)
cat("DoubletFinder classification column:", df_col, "\n")

# Subset the Seurat object to keep only cells classified as "Singlet"
saline2 <- subset(saline2,subset= DF.classifications_0.15_0.005_340 == "Singlet")
head(saline2@meta.data)

#IAV
sweep.res.list.IAV  <- paramSweep(IAV,PCs=1:50,sct=TRUE)
sweep.stats.IAV <- summarizeSweep(sweep.res.list.IAV,GT=FALSE)
bcmvn.IAV <- find.pK(sweep.stats.IAV)
sweep.stats.IAV 
IAV
nExp_poi <- round(0.06152*nrow(IAV@meta.data)) 

# Run DoubletFinder with chosen parameters
IAV <- doubletFinder(IAV, PCs=1:50, pN=0.15, pK=0.005, nExp=nExp_poi, reuse.pANN=FALSE, sct=TRUE) 

# Identify the doublet classification column (its name includes the parameters used)
df_col <- grep("DF.classifications", colnames(IAV@meta.data), value = TRUE)
cat("DoubletFinder classification column:", df_col, "\n")

# Subset the Seurat object to keep only cells classified as "Singlet"
IAV <- subset(IAV,subset= DF.classifications_0.15_0.005_378 == "Singlet")
head(IAV@meta.data) 
```

## Recluster singlet datasets
```{r}
#Saline1
saline1 <- FindNeighbors(saline1, dims = 1:50)
saline1 <- FindClusters(saline1, resolution = 0.5) 
DimPlot(saline1, reduction = "umap", label=TRUE, pt.size = 0.5)

#Saline2
saline2 <- FindNeighbors(saline2, dims = 1:50)
saline2 <- FindClusters(saline2, resolution = 0.5) 
DimPlot(saline2, reduction = "umap", label=TRUE, pt.size = 0.5)

#IAV
IAV <- FindNeighbors(IAV, dims = 1:50)
IAV <- FindClusters(IAV, resolution = 0.5) 
DimPlot(IAV, reduction = "umap", label=TRUE, pt.size = 0.5)
```

## Check Compartments and Feature Counts
```{r QC}
VlnPlot(saline1,features=c("Epcam","Ptprc","Col1a1","Cldn5"),pt.size=0.0)
VlnPlot(saline1,features=c("nFeature_RNA","nCount_RNA","percent.mt"),pt.size=0)

VlnPlot(saline2,features=c("Epcam","Ptprc","Col1a1","Cldn5"),pt.size=0)
VlnPlot(saline2,features=c("nFeature_RNA","nCount_RNA","percent.mt"),pt.size=0)

VlnPlot(IAV,features=c("Epcam","Ptprc","Col1a1","Cldn5"),pt.size=0)
VlnPlot(IAV,features=c("nFeature_RNA","nCount_RNA","percent.mt"),pt.size=0)
```

## Save preprocessing objects
```{r saveRDS}
setwd('/Users/alexanderfoote/R_Projects/scRNA_analysis_7dpi_postIAV/rds')
saveRDS(saline1, "saline1.std_pipeline_preprocessed_singlets.rds")
saveRDS(saline2, "saline2.std_pipeline_preprocessed_singlets.rds")
saveRDS(IAV, "IAV.std_pipeline_preprocessed_IAVseq.rds")
```

### Session information
```{r sessioinfo}
sessionInfo()
```
